### ISSUE TITLE
Creatures leash system not blizzlike

### ISSUE THREAD

**Issue by T1ti** [azerothcore/azerothcore-wotlk#5116]
## EXPECTED BLIZZLIKE BEHAVIOUR:

1. There's a small radius around the aggro point where creature will never leave combat
2. Outside of that radius, the leash is timer based(around ~12sec _(seems to be based on creature level and actually ranges from 11 to 16sec)_ . When that timer expires, the creature resets, as simple as that.
3. several actions will reset that leash timer : 
  -the creature standing still/attacking for 5seconds(not in chase state)
  -the creature being hit by an hostile action from the player : either an attack, or a debuff(even non damaging)
Those actions will just reset the leash timer and the creature will chase for ~12seconds again

## CURRENT BEHAVIOUR:
1. Kind of working, creatures will never reset n a radius around the creature, but see point 2
2. There doesn't seem to be a timer, it looks purely distance based, creatures seem to reset only when they reach a distance from the original aggro location. There is an enormous radius around the original creature location where it will never reset, and the creature just resets when you step outside that radius
3. Player attacks are extending the leash outside the radius a bit currently, by about 10sec. The creature attacking does not extend the leash at all, if you bring the creature out of range and just let it attack you, it will reset after a bit(should not happen)
**-Debuffs don't extend the leash, it has been verified that they do on both Classic AND retail** (debuffs pretty much count as an attack)

## SOURCE:
Based on testing on Classic servers by me and the vmangos project.
![image](https://user-images.githubusercontent.com/40864460/113485279-017aa500-94ad-11eb-884f-cee5957cd466.png)
Made a drawing to show a bit how it works, the sizes are guessed
edit : new source : blizard's official creature tool from 2009, I believe the "timeout" is actually the leash timer, it's set to 0 for instance creatures (unlimited)
![image](https://user-images.githubusercontent.com/40864460/116008968-3fd03380-a617-11eb-99f2-68d87ea81347.png)
I described those settigns here : https://github.com/azerothcore/azerothcore-wotlk/issues/5455 

## STEPS TO REPRODUCE THE PROBLEM:

1. just test the points

## AC HASH/COMMIT:
4b7454574412cfe23bb5902b9b3d42d2313b3541

**Comment by T1ti** [comment]
Note : the current AC system feels extremly obnoxious in cavesa nd tunnels with turns thata ren't a straight line, i had creatures chasing me for nearly 40 seconds in the jasperlode mine without hitting them because I was theorycally not existing the radius

**Comment by T1ti** [comment]
https://www.youtube.com/watch?v=x6P801lzTOk
Video example, this kobold keeps chasing for 50 seconds without hitting him in a huge circle

**Comment by T1ti** [comment]
Is anyone interested in fixing this ? THe leash system has to be one of the worst part of AC, it makes caves unplayable

**Comment by benjymansy123** [comment]
https://github.com/chromiecraft/chromiecraft/issues/3382
Related to this, also includes a writeup about [Classic](https://classicwow.live/guides/793/leash-behavior?g=classic).

**Comment by Maelthyrr** [comment]
Related to UN comment, need to be reconfirm with woltk classic. https://github.com/azerothcore/azerothcore-wotlk/pull/4897

**Comment by T1ti** [comment]
It's still how it works on retail, it has never changed

**Comment by heyitsbench** [comment]
Worth adding as it's in the other (closed) issue but not this one, the vmangos implementation commit can be found [here](https://github.com/vmangos/core/commit/7d2f1e2923734ea1b983eeaa0208cfd158d4af02).

**Comment by neifion-00000000** [comment]
https://github.com/Rebirth-MaNGOS/MaNGOS  we had working leashing fyi, it wasn't perfect but after a dozen or so fixes it was pretty good.  There is no inclusion of some of the other values mentioned but it would be a very solid starting point if someone is capable of using andkems work.

**Comment by T1ti** [comment]
It's probably better to look at the vmangos implementation

**Comment by Maelthyrr** [comment]
Someone can check if this work always ? https://github.com/azerothcore/azerothcore-wotlk/commit/fdc29c7a0277a848dc3a0279367a7e1d03b1b8bd

**Comment by neifion-00000000** [comment]
https://youtu.be/uGW6hdnuYgk  just further evidence, this issue is correct, if vmangos is in fact running on a 12-15sec timer with a mobile radius then that would likely be the one we want to implement

**Comment by T1ti** [comment]
https://web.archive.org/web/20210305073636/https://forum.classicwow.live/topic/625/leash-behavior
here's some extra research on leash duration

**Comment by heyitsbench** [comment]
> here's some extra research on leash duration

There is also [this YouTube video](https://www.youtube.com/watch?v=xnrUkbEbiJk) discussing the subject, released late March this year, geared more as a PSA towards hardcore players. Unsurprisingly, it very much seems to be the same system as Wrath Classic in my limited experiencew.

**Comment by doowahree** [comment]
I think to implement this, we can approach with the simplified leashing model to start with.

Unit will continue aggro to player if...

With the following info:
```
1. Whether the player is in boundary.
2. Remaining leash timer.
3. Last attack position.
```

and ANY of the following are true...
```
1. Creature is within the bounding room. (Already does this).
2. When the leash timer > 0. (Not yet implemented)
    1. Leash timer only decreases if creature not in bounding room
    2. Leash timer resets if...
        1. Any hostile action is done against it.
        2. The creature's attack position is same between last and this time's position.
```

Then continue aggro.

Implementation pathway:

**Checkpoint 1 - timed leash outside of bounding box**

1. Add `m_remainingLeashTime` to `Creature`
2. Rename `CreatureAI::CheckInRoom()` -> `CreatureAI::CheckLeash(uint32 diff)` with default diff of 0.
3. In `CreatureAI::CheckLeash` if not in boundary instead of directly entering evade mode...
    1. Set m_remainingLeashTime -= diff
    2. If m_remainingLeashTime < 0, leash (enter evade mode) and reset `m_remainingLeashTime`. Otherwise let aggro.

Issues:
1. The diffs will always be >2.5s, which is fine for the simplified version imo.

**Checkpoint 2 - reset leash upon hostile action**

1. In `CreatureAI::AttackedBy`, reset the creature's `m_remainingLeashTime`.

**Checkpoint 3 - Reset upon melee attack**

1. In `Unit`, add `m_lastAttackPosition`
1. In `Unit::Attack`, there's already conversion to see if unit is `Creature`. If so we can reset the Creature's `m_remainingLeashTime` if `m_lastAttackPosition` and `this` (`Unit` itself is a child of Position struct) distance is a small amount.

---------------------

Once this is in, other folks can make this more fancy or whatnot but this seems like a good start.


If this look good... If anyone wants to implement this, please do so, if not I can take a stab at it soon.

**Comment by doowahree** [comment]
Hmmm, so there is an issue with using `CreatureAI::AttackedBy`. So far it seems to only be called in initial aggro - https://github.com/azerothcore/azerothcore-wotlk/blob/3ff8de2086a2f32316ec0041963968c6eccf8958/src/server/game/Entities/Unit/Unit.cpp IIUC...

So leash extension can't be done by this. The alternative to simplify things is to do so by when unit is dealing damage, check if the unit is creature and if so, get the AI and reset the leash. That may be an okay alternative to start with.

**Comment by Dezaimasu** [comment]
I assume this problem still exists. Posting here since there's already enough duplicate tickets related to this issue. 
After dying few times to a murlock I did a quick test, and here's my experience. 
Basically as long as the mob can hit you consistently, it won't de-aggro from you, ever. It's easily noticeable with murlocks since they're fast. I run all the way from the the Tirisfal shore to the Brill with mob hitting me all the time. Other mobs won't de-aggro as well if you let them hit you consistently, as long as they can't hit you for some time, they will de-aggro eventually.

**Commit**
2ebbd5648857d3872bcf95995c0f1b1f1ecdd2cd (latest as of today)

**Modules**
mod-ah-bot
mod-autobalance
mod-transmog

**Expected Blizzlike behavior**
I haven't played in a while, but I believe at least in WotLK there was some "hard limit" for the range indicating how far you can kite the non-dungeon mob away from it's initial position, after that the mob will instantly evade no matter what.

**Steps to reproduce the problem**
1 `.cheat god`
2 `.go zonexy 61 29 85`
3 Aggro "Vile Fin Puddlejumper", let it get close to you, and then try to run away from it.

**Video**
https://files.catbox.moe/sdhhps.mp4

![World of Warcraft 2024-10-23 14-40-39 mp4_snapshot_02 09 027](https://github.com/user-attachments/assets/f46e3418-1df1-4019-9606-275777e5018f)

**Comment by nyilas7** [comment]
Hello, please consider the hard-limit solution, mentioned by Dezaimasu (a hard-coded distance boundary inside the leash extension check).

Why this is better for gameplay:
1. It prevents players from abusing caster pathing and safely drag high-level elite mobs.
2. By scaling the hard limit, players still have plenty tactical space to legitimately kite mobs within a local area. It only cuts off broken/endless chases.
3. Infinite chasing forces the server to continually calculate pathfinding grids.
4. Preserving standard instance mechanics, meaning dungeon and raid mobs will still hunt players endlessly as intended.
5. Hardcore players will stop tearing out their hair by dying to casters infinitely chasing them.
6. Less frustrating death in caves.

I've applied these modifications locally and verified the fix in multiple test scenarios. It looks good to me, but feel free to double-check it.

src/server/game/Entities/Creature/Creature.cpp

From line 2700:

        // pussywizard: don't check distance to home position if recently damaged (allow kiting away from spawnpoint!)
        // xinef: this should include taunt auras
        if (!isWorldBoss() && (GetLastLeashExtensionTime() + GetLeashTimer() > GameTime::GetGameTime().count() || HasTauntAura()))
        {
        // Hard limit: Prevents infinite spellcaster kiting over massive distances
        float maxHardLeash = sWorld->getFloatConfig(CONFIG_CREATURE_LEASH_RADIUS) * 2.4f;
    
        if (IsInDist2d(&m_homePosition, maxHardLeash))
                return true;
        else
                return false; // Force the creature to stop attacking and evade if it goes too far!
        }

**Comment by wetbrownsauce** [comment]
If the suggestion isn't how 3.3.5a leashing functions, then it shouldn't be added. Classic didn't have a max leash.

**Comment by nyilas7** [comment]
And Classic also didn't have caster mobs chasing you across the map.

The current way it functions is also not wotlk3.3.5a.

**Comment by wetbrownsauce** [comment]
Right, but the fix for something not being blizzlike is not to implement something else that isn't blizzlike. Leashing is a complicated thing to nail down, and adding hard caps isn't fixing anything, it is limiting the system from its original design to avoid an issue that hasn't been corrected. That's just a hackfix

**Comment by nyilas7** [comment]
You can call it hackfix or workaround, it solves a problem many player faces. And the average player doesn't care about whether it is "blizzlike" or not, they want to enjoy the game and not hit their head on the wall every time they meet a caster mob.

You currently don't have a solution that is not hackfix, and the gameplay is better with this workaround than without it.

**Comment by wetbrownsauce** [comment]
Right, but again, you're suggesting it be merged into the core, which is trying to eliminate hackfixes. You can use whatever bandage fixes for problems you want, and if someone else wants to, they can go ahead, but it shouldn't be included in azerothcore.

**Comment by nyilas7** [comment]
Reverse-engineering is already a hackfix, it is never going to be the same as the original server code.

I agree with what you are saying, a solution would be better. But at the moment nobody has it, and it is mid 2026, for a ticket that is open since Apr 3 2021...

Also, I am not so sure that Blizzard doesn't have a hard limit in their code. How do you know?

Because someone wrote some articles or made some videos on the internet, that also does not qualify as blizzlike.

**Comment by wetbrownsauce** [comment]
While the system isn't blizzlike yet, you can see from the commit history that leashing has been a constant effort to get right. There have been many changes and updates to the leashing system. 

Blizzlike behavior does not mean blizzlike code. It will never be the original server code, that is an absurd concept. 

You mention that this issue has been up since 2021 with no fix, but the issue is with the exact fix you're describing. This issue was that hardcoded reset distances are not blizzlike and should be removed.

**Comment by nyilas7** [comment]
Nobody doubted the efforts to make the leashing better.

"Blizzlike behavior does not mean blizzlike code. It will never be the original server code; that is an absurd concept."

We agree. That is why some flexibility helps at times.

"You mention that this issue has been open since 2021 with no fix, but the issue is with the exact fix you're describing. This issue was that hardcoded reset distances are not blizzlike and should be removed."

Today the issue is rather mobs that can cast spells chase you across ridiculous distances.

You don't want to workaround it, fine (yet again, I am not 100% sure there is no hard limit on Blizzard Classic servers).

If we are at open-source development, to take an example, Linus Torvalds says often, never break the userspace.

Today, the "userspace" is broken, this is a major pain-point bug without a solution for years, despite the efforts.

**Comment by wetbrownsauce** [comment]
Can you provide a video of this bug you're describing? Casters are resetting fine for me, although my core isn't up to date so perhaps a recent change broke caster leashing.

**Comment by nyilas7** [comment]
Here is a video that i've recorded:

https://youtu.be/At8EOnG5G_k

It must be like that since minimum 1.5 months.

**Comment by wetbrownsauce** [comment]
yeah that doesn't happen for me, I just tried with the same mobs in the same spot. My core is from august 2025 so it must have been a change to movement or leashing that happened sometime between then and now. Sounds like a regression somewhere... Idk maybe it would be best to open a new issue about that specifically.

**Comment by heyitsbench** [comment]
> please consider the hard-limit solution, mentioned by Dezaimasu (a hard-coded distance boundary inside the leash extension check).

This will not be added to AzerothCore.

**Comment by nyilas7** [comment]
> > please consider the hard-limit solution, mentioned by Dezaimasu (a hard-coded distance boundary inside the leash extension check).
> 
> This will not be added to AzerothCore.

I guess I understood the first time, thank you.

A fix on the caster mob leashing problem, that is considered acceptable by your team, will it be added to AzerothCore?

**Comment by heyitsbench** [comment]
A solution that's accurate to official servers would almost *always* be accepted.

EDIT: Research is currently being done to try and further iron out the behavior you're talking about on official servers.

**Comment by nyilas7** [comment]
Sorry for taking my time and look it up what could be the problem, only to produce that _inaccurate_ solution with my limited knowledge on the subject. I should have put more time in it.

Since this was working 2025. August (apparently), I guess, some solutions that were not accurate for official servers had been accepted as well unfortunately (hence the bug?).

https://github.com/chromiecraft/chromiecraft/issues/3382

Should I open an another ticket on ChromieCraft bug tracker or not needed?

**Comment by heyitsbench** [comment]
The issue is still open, so another issue is not needed.

**Comment by nyilas7** [comment]
I might be wrong so please correct me.

I think the problem might be with 

src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp

around line :180

        _lastTargetPosition.reset();
        if (cOwner)
        {
            if (isStoppedBecauseOfCasting)
            {
                // Don't reset leash timer if it's a spell like Shoot with a short cast time.
                /// @todo: Research how it should actually work.
                Spell *spell = cOwner->GetFirstCurrentCastingSpell();
                bool spellHasLongCast = spell && spell->GetCastTime() > 1 * SECOND * IN_MILLISECONDS;
                if (spellHasLongCast)
                    cOwner->UpdateLeashExtensionTime();
            }
            else
                cOwner->UpdateLeashExtensionTime();

            cOwner->SetCannotReachTarget();
        }
        return true;

This reset the leash clock to now on every single cast? Since casters cast repeatedly, the clock was constantly being reset?

By changing the code to:

```
   _lastTargetPosition.reset();
        if (cOwner)
        {
            if (!isStoppedBecauseOfCasting)
                cOwner->UpdateLeashExtensionTime();

            cOwner->SetCannotReachTarget();
        }
        return true;
```
  
Every time a caster mob stops to cast a spell, the leash timer now ticks down instead of resetting. Once the timer expires, the mob evades and walks home.

I've compiled and tested, works for me.

**Comment by nyilas7** [comment]
Here is the commit, before 2025. July 29, these lines were not present in the code:

https://github.com/azerothcore/azerothcore-wotlk/commit/67aa022dbf4c303aa7457b10b6c9dee9049b0d67

**Comment by wetbrownsauce** [comment]
There would have to be more to the picture, since my august core has that commit with no issue. I still would think that something happened somewhere else in the core post august that must have altered how that code block works, idk.

**Comment by nyilas7** [comment]
Which revision/build of the server are you on?

**Comment by wetbrownsauce** [comment]
I'm on 73317b2

**Comment by nyilas7** [comment]
It seems that the bug is present on that version of AC as well.

Sorry for the video quality: https://www.youtube.com/watch?v=hAUQJCytT-c

**Comment by wetbrownsauce** [comment]
Interesting, it didn't happen with the murlocs for me, but yeah I have it happen with those mobs if they're allowed to stop and cast. Resetting them requires getting so much distance that they can't ever stop to try and cast, but that definitely isn't right. 

Seems like that todo in the code needs to actually be handled then
